flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/ForStGeneralMultiGetOperation.java

```diff
@@ -33,39 +37,169 @@
+                            for (int i = 0; i < getRequests.size(); i++) {
+                                completeExceptionallyRequest(
+                                        getRequests.get(i), "Error occurred when multiGet", e);
+                            }
+                        }
+                        if (error.get() != null) {
+                            if (subProcessFinished != null) {
+                                subProcessFinished.run();
+                            }
+                            return;
+                        }
+                        for (int i = 0; i < getRequests.size(); i++) {
+                            ForStDBGetRequest<?, ?, ?, ?> request = getRequests.get(i);
+                            try {
+                                if (error.get() == null) {
+                                    request.completeStateFuture(values.get(i));
+                                } else {
+                                    completeExceptionallyRequest(
+                                            request,
+                                            "Error already occurred in other state request of the same "
+                                                    + "group, failed the state request directly",
+                                            error.get());
+                                }
+                            } catch (Exception e) {
+                                error.set(e);
+                                completeExceptionallyRequest(
+                                        request, "Error when complete get future.", e);
+                                future.completeExceptionally(e);
+                            }
+                        }
+
+                        if (counter.addAndGet(-getRequests.size()) == 0
+                                && !future.isCompletedExceptionally()) {
+                            future.complete(null);
+                        }
+                        if (subProcessFinished != null) {
+                            subProcessFinished.run();
+                        }
+                    });
+        }
+        return future;
+    }
+
+    private void completeExceptionallyRequest(
+            ForStDBGetRequest<?, ?, ?, ?> request, String message, Exception e) {
+        request.completeStateFutureExceptionally(message, e);
+    }
+
+    private void classifyAndSplitRequests(
+            List<List<ForStDBGetRequest<?, ?, ?, ?>>> splitRequests,
+            List<ForStDBGetRequest<?, ?, ?, ?>> mapCheckRequests) {
+        List<ForStDBGetRequest<?, ?, ?, ?>> getRequests = new ArrayList<>();
         for (int i = 0; i < batchRequest.size(); i++) {
             ForStDBGetRequest<?, ?, ?, ?> request = batchRequest.get(i);
+            if (request instanceof ForStDBMapCheckRequest) {
+                mapCheckRequests.add(request);
+            } else {
+                getRequests.add(request);
```

Comment:
How about using round-robin to add request into `splitRequests`?

Reply:
I'm worried that this will break the locality of access, now adjacent requests are in a small batch.